### PR TITLE
[Bugfix][KV Offloading] Defer stores until block IDs are tracked

### PR DESCRIPTION
**Pull request by yatesdr**
# `[Bugfix][KV Offloading] Defer stores until block IDs are tracked`

## Purpose

Fix a fatal `OffloadingConnector._build_store_jobs` assertion when KV pressure
causes eviction/preemption while tiered offloading and async scheduling are
active.

The store path currently derives `num_chunks` from token progress, then assumes
that two independently maintained collections have reached the same boundary:

- content-hash-derived `offload_keys`; and
- scheduler-propagated physical GPU `block_ids`.

Eviction/preemption can expose a transient state where the token-derived store
boundary is ahead of one of those collections. The resulting slices have
different lengths, and this assertion terminates `EngineCore`:

```python
assert len(offload_keys) == len(offload_block_ids)
```

This change computes a ready boundary independently for every KV group:

```python
min(
    num_storable_chunks,
    len(group_state.offload_keys),
    len(group_state.block_ids) // blocks_per_chunk,
)
```

Only that common ready prefix is offered to the offload manager. Store progress
also advances only through the capped boundary, so a lagging chunk is retried
after its key and complete set of source block IDs are tracked. This is
important: merely removing the assertion and relying on `zip()` would avoid the
crash but could advance the cursor past an unstored chunk, silently losing it.

When all three state sources are synchronized, the ready boundary equals the
existing token-derived boundary and behavior is unchanged.

## Production evidence

The failure was reproduced on a GLM-5.2 hybrid deployment using TP4/DCP4,
MTP3, async scheduling, `TieringOffloadingSpec`, and a 644,864-token GPU KV
pool:

- concurrency 16 x 50k context (~800k active tokens): fatal assertion during
  eviction;
- concurrency 32 sweep: same assertion;
- concurrency 16 x 16k and concurrency <= 8 x 50k: no assertion while active
  demand remained below the pool;
- BF16, E4M3, and block-INT8 collective modes all reproduced the failure,
  ruling out the collective wire codec.

The deployed downstream vLLM commit is `7ea567a2458a4800a6a0e3e0a6ba41fcbd00d146`.
The complete patch applies cleanly to that commit as well as current upstream
main.

## Test Plan

1. Reproduce the lag directly in the existing OffloadingConnector scheduler
   fixture: make three chunks storable while only two GPU block IDs are tracked.
2. Verify the first store job contains the two ready chunks and advances only
   to chunk two.
3. Add the third block ID and verify a second store job retries exactly chunk
   three.
4. Run the complete neighboring scheduler suite, including async/sync
   scheduling, request finish, preemption, reset, MTP/EAGLE, hybrid groups, and
   sliding-window cases.
5. Run all repository pre-commit hooks applicable to the changed files.
6. Before submission, repeat the original above-pool workload on the target
   deployment and verify zero assertion, zero `EngineDeadError`, no restart,
   and successful request completion.

## Test Result

Completed locally:

```text
pytest tests/v1/kv_connector/unit/offloading_connector/test_scheduler.py
92 passed, 15 warnings in 40.95s
```

```text
pre-commit run --files <scheduler.py> <test_scheduler.py>
all applicable hooks passed
```

- `git diff --check`: passed.
- Patch applicability against deployed v19 commit `7ea567a`: passed.
- The new regression passes and proves the deferred chunk is retried rather
  than skipped.

Target-box overflow reproduction:

- Patched image SHA-256 began with `41078d5f`; the mounted scheduler SHA-256
  began with `a68a28e5`.
- Deployment: block-INT8 wire mode, offloading enabled with 64 GB host DRAM,
  480k maximum model length, 16 maximum sequences, 0.970 GPU memory
  utilization, TP4/DCP4/MTP3, and a 611,840-token boot KV pool.
- 16/16 unique-prefix requests completed successfully with
  `finish_reason=stop`; each carried approximately 49,800 prompt tokens.
- Active demand reached approximately 797k tokens against a 611,840-token GPU
  KV pool, forcing the eviction path that crashed the unpatched connector.
- Prefix-cache metrics recorded 797,218 queries and zero hits, confirming that
  the requests were freshly prefilled rather than cache-served.
- `RestartCount` remained 0 -> 0 under a one-second watcher; container identity
  and `StartedAt` remained unchanged throughout the 11-minute run.
- No `_build_store_jobs` assertion, `EngineDeadError`, OOM, HTTP 5xx, or request
  error occurred.
- A post-load liveness request returned the expected answer `4` from the same
  engine instance. The container ID, `StartedAt`, and `RestartCount=0` remained
  unchanged.

All production acceptance criteria passed.

No model evaluation is required for this change because it does not alter model
weights, kernels, logits, sampling, or transferred KV values. It changes only
when scheduler metadata declares a store chunk ready. The target overflow gate
above remains required because it validates the production scheduler lifecycle.

## Non-duplication

Open-PR searches and fetched git-history searches for the assertion, store
boundary, and preemption/block-ID terms found no existing fix as of 2026-07-19.

PR [#48596](https://github.com/vllm-project/vllm/pull/48596) is related but not
duplicative: it fixes final-block storage and a block-reuse race at request
finish. Both the deployed commit and current upstream main include that change
while retaining the failing length assertion addressed here.

## AI assistance disclosure

OpenAI Codex assisted with failure analysis, patch preparation, and regression
test development. The commit contains an `Assisted-by` trailer. The production
overflow validation remains a merge gate and its results will be added here.

### CODE REVIEW

**Comment by yatesdr** [comment]
Feedback please on no merge close?

**Comment by voipmonitor** [comment]
Closing this PR without a published replacement was a process mistake.

What happened:

- #133 targeted `fix/gg-rebased-cutlass453-sm120-20260718`, a transient integration branch, rather than the canonical `dev/gilded-gnosis` branch.
- During the later PR-topology cleanup, branches that were stacked on obsolete integration bases were closed so that fixes could be represented as direct-GG PRs.
- An exact local port onto GG was prepared, but it was not pushed and no replacement PR was opened before #133 was closed. The existence of a local branch was incorrectly treated as completion of the migration.

Resolution:

- Replacement: #153
- #153 is directly based on current `dev/gilded-gnosis`.
- The fix was re-audited against upstream vLLM #49146 and adjusted to keep the readiness invariant inside `storable_chunks()`.
- It covers both independently lagging metadata sources (`offload_keys` and complete `block_ids` chunks), retains the consistency assertion, and adds queued-abort coverage.
- The full offloading-connector unit suite passes: 142 tests.

Prevention for future PR consolidation:

1. Verify `base.ref` and merge-base against the canonical branch before publishing.
2. Open, validate, and link the replacement PR before closing the old PR.
3. Track every migration as `old PR -> replacement PR -> tested commit`; a local branch is not a published replacement.
4. Never close a superseded fix without a comment linking its replacement or explicitly stating that the patch was intentionally dropped.

No implementation from #133 is intentionally being discarded; #153 is its cleaner and more thoroughly validated successor.
